worker/caasfirewallerembedded/applicationworker.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasfirewallerembedded
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/juju/charm/v8"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v2"
+	"github.com/juju/worker/v2/catacomb"
+
+	"github.com/juju/juju/caas"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/core/watcher"
+)
+
+type applicationWorker struct {
+	catacomb       catacomb.Catacomb
+	controllerUUID string
+	modelUUID      string
+	appName        string
+
+	firewallerAPI CAASFirewallerAPI
+
+	broker         CAASBroker
+	portMutator    PortMutator
+	serviceUpdater ServiceUpdater
+
+	appWatcher   watcher.NotifyWatcher
+	portsWatcher watcher.StringsWatcher
+
+	lifeGetter LifeGetter
+
+	initial           bool
+	previouslyExposed bool
+
+	currentPorts portRanges
+
+	logger Logger
+}
+
+type portRanges map[network.PortRange]bool
+
+func newPortRanges(in []network.PortRange) portRanges {
+	out := make(portRanges)
+	for _, p := range in {
+		out[p] = true
+	}
+	return out
+}
+
+func (pg portRanges) equal(in portRanges) bool {
+	if len(pg) != len(in) {
+		return false
+	}
+	return reflect.DeepEqual(pg, in)
+}
+
+func (pg portRanges) toServicePorts() []caas.ServicePort {
+	out := make([]caas.ServicePort, len(pg))
+	for p := range pg {
+		out = append(out, caas.ServicePort{
+			// TODO(embedded): add name to `network.PortRange`?
+			// Name:       p.Name,
```

Comment:
I think name is needed by k8s if there's more than one service port defined?

Reply:
yeah, so I left a TODO for later changes.